src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultTagHelperCompletionService.cs

```diff
@@ -0,0 +1,299 @@
+                attributes,
+                ancestorTagName,
+                ancestorIsTagHelper,
+                HtmlSchemaTagNames.Contains);
+
+            var completionItems = new List<CompletionItem>();
+            var completionResult = _razorTagHelperCompletionService.GetElementCompletions(elementCompletionContext);
+            foreach (var completion in completionResult.Completions)
+            {
+                var razorCompletionItem = new CompletionItem()
+                {
+                    Label = completion.Key,
+                    InsertText = completion.Key,
+                    FilterText = completion.Key,
+                    SortText = completion.Key,
+                    Kind = CompletionItemKind.TypeParameter,
+                    CommitCharacters = ElementCommitCharacters,
+                };
+
+                completionItems.Add(razorCompletionItem);
+            }
+
+            return completionItems;
+        }
+
+        // Internal for testing
+        internal static IEnumerable<KeyValuePair<string, string>> StringifyAttributes(SyntaxList<RazorSyntaxNode> attributes)
+        {
+            var stringifiedAttributes = new List<KeyValuePair<string, string>>();
+
+            for (var i = 0; i < attributes.Count; i++)
+            {
+                var attribute = attributes[i];
+                if (attribute is MarkupTagHelperAttributeSyntax tagHelperAttribute)
+                {
+                    var name = tagHelperAttribute.Name.GetContent();
+                    var value = tagHelperAttribute.Value.GetContent();
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, value));
+                }
+                else if (attribute is MarkupMinimizedTagHelperAttributeSyntax minimizedTagHelperAttribute)
+                {
+                    var name = minimizedTagHelperAttribute.Name.GetContent();
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, string.Empty));
+                }
+                else if (attribute is MarkupAttributeBlockSyntax markupAttribute)
+                {
+                    var name = markupAttribute.Name.GetContent();
+                    var value = markupAttribute.Value.GetContent();
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, value));
+                }
+                else if (attribute is MarkupMinimizedAttributeBlockSyntax minimizedMarkupAttribute)
+                {
+                    var name = minimizedMarkupAttribute.Name.GetContent();
+                    stringifiedAttributes.Add(new KeyValuePair<string, string>(name, string.Empty));
+                }
+            }
+
+            return stringifiedAttributes;
+        }
+
+        private static (string ancestorTagName, bool ancestorIsTagHelper) GetNearestAncestorTagInfo(IEnumerable<SyntaxNode> ancestors)
+        {
+            foreach (var ancestor in ancestors)
```

Comment:
Spoke offline. The ancestor of a start tag can't be another start tag (Except in malformed cases). Should check for `MarkupElement` or `MarkupTaghelperElement` and get its `StartTag` property.